metaflow/plugins/kfp/kfp_step_function.py

```diff
@@ -1,10 +1,242 @@
+        "--max-user-code-retries %d" % user_code_retries,
+        (
+            "--input-paths %s" % start_task_id_params_path
+            if node_name == "start"
+            else f"--input-paths ${INPUT_PATHS_ENV_NAME}"
+        ),
+    ]
+
+    if need_split_index:
+        step.append(f"--split-index ${SPLIT_INDEX_ENV_NAME}")
+
+    step.extend(tags_extended)
+
+    if namespace:
+        step.append("--namespace %s" % namespace)
+
+    cmds.append(" ".join(entrypoint + top_level + step))
+    step_cli_string = " && ".join(cmds)
+    return step_cli_string
+
+
+def _command(
+    cd_into_metaflow_package_cmd: str,
+    clean_volume_cmd: str,
+    step_cli: List[str],
+    task_id_template: str,
+    step_name: str,
+    flow_name: str,
+) -> str:
+    """
+    Analogous to batch.py
+    """
+    retry_count_python = (
+        "import os;"
+        'name = os.environ.get("MF_ARGO_NODE_NAME");'
+        'index = name.rfind("(");'
+        'retry_count = (0 if index == -1 else name[index + 1: -1]) if name.endswith(")") else 0;'
+        "print(str(retry_count))"
+    )
+
+    mflog_expr = export_mflog_env_vars(
+        flow_name=flow_name,
+        run_id="{run_id}",
+        step_name=step_name,
+        task_id=task_id_template,
+        retry_count=f"`python -c '{retry_count_python}'`",
+        datastore_type="s3",
+        datastore_root="$METAFLOW_DATASTORE_SYSROOT_S3",
+        stdout_path=STDOUT_PATH,
+        stderr_path=STDERR_PATH,
+    )
+
+    step_cmds = []
+    step_cmds.append("echo 'Task is starting.'")
+    step_cmds.extend(step_cli)
+
+    step_expr = bash_capture_logs(" && ".join(step_cmds))
+
+    # construct an entry point that
+    # 1) Clean attached volume if any
+    # 2) Initializes the mflog environment (mflog_expr)
+    # 3) Bootstraps a metaflow environment (cd_into_metaflow_package_cmd)
+    # 4) Executes a task (step_expr)
+    cmd_str = (
+        f"{clean_volume_cmd} "
+        f"&& mkdir -p {LOGS_DIR} && {mflog_expr} "
+        f" {cd_into_metaflow_package_cmd} "
+        f"&& {step_expr};"
+    )
+
+    # after the task has finished, we save its exit code (fail/success)
+    # and persist the final logs. The whole entrypoint should exit
+    # with the exit code (c) of the task.
+    #
+    # Note that if step_expr OOMs, this tail expression is never executed.
+    # We lose the last logs in this scenario.
+    cmd_str += "c=$?; %s; exit $c" % BASH_SAVE_LOGS
+    return cmd_str
+
 
 def kfp_step_function(
-    cmd_template: str,
     metaflow_run_id: str,
     metaflow_configs: Dict[str, str],
+    cd_into_metaflow_package_cmd: str,
+    clean_volume_cmd: str,
+    task_id: str,
+    task_id_template: str,
+    step_name: str,
+    flow_name: str,
+    namespace: str,
+    tags: List[str],
+    need_split_index: bool,
+    environment_type: str,
+    logger_type: str,
+    monitor_type: str,
+    user_code_retries: int,
+    workflow_name: str,
+    script_name: str,
```

Comment:
This functions returns type `object` is that still necessary?

Must it return this complicated tuple still?

```python
    ret = namedtuple(
        "StepOpRet", ["foreach_splits"] + list(preceding_component_inputs_dict.keys())
    )(*values)
    return ret
```

Reply:
Good point, we can return a `List[str]`.